test/umschreiben_clj/requires_test.clj

```diff
@@ -0,0 +1,273 @@
+                             'x.y.z))))))
+
+  (deftest known-issues
+    ; This can't happen actually, it could if we support replacing refers, for that we should provide an extra bit of information
+    ; like this {'myfn 'myfn-shadow}.
+    (testing "warning: it is possible to introduce shadowing"
+      ;; to fix this we would have to collect scope as we traverse
+      (is (match? '(do (ns aa (:require [x.y.z :refer [myfn-shadow]])) (let [myfn-shadow 1] (myfn 1)))
+                  (n/sexpr (requires/replace
+                             (code->node '(ns aa (:require [a.b.c :refer [myfn]]))
+                                         '(let [myfn-shadow inc] (myfn 1)))
+                             'a.b.c
+                             '[x.y.z :refer [myfn-shadow]]))))))
+
+  (testing "don't replace :refer when shadowed by (let) binding"
+    (is (match? '(do (ns aa (:require [x.y.z :refer [myfn-shadow]])) (let [myfn-shadow 1] (inc myfn-shadow)))
+                (n/sexpr (requires/replace
+                           (code->node '(ns aa (:require [a.b.c :refer [myfn]]))
+                                       '(let [myfn 1] (inc myfn)))
+                           'a.b.c
+                           '[x.y.z :refer [myfn-shadow]])))))
+
+  (testing "don't replace :refer when shadowed by (fn) binding"
+    (is (match? '(do (ns aa (:require [x.y.z :as z])) (fn [myfn] (inc myfn)))
+                (n/sexpr (requires/replace
+                           (code->node '(ns aa (:require [a.b.c :refer [myfn]]))
+                                       '(fn [myfn] (inc myfn)))
+                           'a.b.c
+                           '[x.y.z :as z]))))))
+
+(deftest add-test
+  (testing "effect on 'ns' code"
+    (are [current new result]
+         (match? (strictly result)
+                 (n/sexpr (requires/add (code->node current) new)))
+
+         '(ns aa (:require [a.b.c]))  'x.y.z  '(ns aa (:require x.y.z [a.b.c]))
+         '(ns aa (:require [x.y.z]))  'x.y.z  '(ns aa (:require x.y.z))
+         '(ns aa (:using a.b.c))      'x.y.z  '(ns aa (:using a.b.c) (:require x.y.z))
+         '(ns aa)                     'x.y.z  '(ns aa (:require x.y.z))
+         '(ns aa (:using e.f.g))      'x.y.z  '(ns aa (:using e.f.g)(:require x.y.z))
+         '(ns aa (:using a.b.c))      'x.y.z  '(ns aa (:using a.b.c)(:require x.y.z))))
+
+  (testing "if the namespace is already required behaves like `replace`"
+    (let [namespace (code->node '(ns aa (:require [a.b.c :as c])) '(def bb (c/myfn 2)))]
+      (is (match? '(do (ns aa (:require [a.b.c :refer [myfn]])) (def bb (myfn 2)))
+                  (n/sexpr (requires/add namespace '[a.b.c :refer [myfn]]))))
+      (is (= (n/sexpr (requires/replace namespace 'a.b.c '[a.b.c :refer [myfn]]))
+             (n/sexpr (requires/add namespace '[a.b.c :refer [myfn]]))))))
+
+  ;; I don't find a use case for this! unless we want to provide low-level functions that should be composed. But
```

Comment:
This is from me, let's remove the comment and/or the test. Do you think the test makes sense?